src/Native/Unix/System.Native/pal_networking.c

```diff
@@ -270,64 +297,123 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+    result = gethostname((char*)name, _POSIX_HOST_NAME_MAX);
+    bool includeIPv4Loopback = true;
+    bool includeIPv6Loopback = true;
 
-    for (struct addrinfo* ai = *info; ai != NULL; ai = ai->ai_next)
+    struct ifaddrs* addrs = NULL;
+    if (result == 0 && strcasecmp((const char*)address, name) == 0)
     {
-        switch (ai->ai_family)
+        // Get all interface addresses if the host name corresponds to the local host.
+        result = getifaddrs(&addrs);
+
+        // If getifaddrs fails, just skip it, the data are not crucial for the result.
+        if (result == 0)
         {
-            case AF_INET:
+            // Count the number of IP end points.
+            for (struct ifaddrs* ifa = addrs; ifa != NULL; ifa = ifa->ifa_next)
             {
-                struct sockaddr_in* inetSockAddr = (struct sockaddr_in*)ai->ai_addr;
+                if (ifa->ifa_addr == NULL)
+                {
+                    continue;
+                }
 
-                ConvertInAddrToByteArray(endPoint->Address, NUM_BYTES_IN_IPV4_ADDRESS, &inetSockAddr->sin_addr);
-                endPoint->IsIPv6 = 0;
-                break;
-            }
+                // Skip the interface if it isn't UP.
+                if ((ifa->ifa_flags & IFF_UP) == 0)
+                {
+                    continue;
+                }
 
-            case AF_INET6:
-            {
-                struct sockaddr_in6* inet6SockAddr = (struct sockaddr_in6*)ai->ai_addr;
+                if (ifa->ifa_addr->sa_family == AF_INET)
+                {
+                    // Remember if there's at least one non-loopback address for IPv4, so that they will be skipped.
+                    if ((ifa->ifa_flags & IFF_LOOPBACK) == 0)
+                    {
+                        includeIPv4Loopback = false;
+                    }
 
-                ConvertIn6AddrToByteArray(endPoint->Address, NUM_BYTES_IN_IPV6_ADDRESS, &inet6SockAddr->sin6_addr);
-                endPoint->IsIPv6 = 1;
-                endPoint->ScopeId = inet6SockAddr->sin6_scope_id;
-                break;
+                    entry->IPAddressCount++;
+                }
+
+                if (ifa->ifa_addr->sa_family == AF_INET6)
+                {
+                    // Remember if there's at least one non-loopback address for IPv6, so that they will be skipped.
+                    if ((ifa->ifa_flags & IFF_LOOPBACK) == 0)
+                    {
+                        includeIPv6Loopback = false;
+                    }
+
+                    entry->IPAddressCount++;
+                }
             }
+        }
+    }
+#endif
 
-            default:
-                // Skip non-IPv4 and non-IPv6 addresses
-                continue;
+    if (entry->IPAddressCount > 0)
+    {
+        entry->IPAddressList = (IPAddress*)calloc((size_t)entry->IPAddressCount, sizeof(IPAddress));
+        if (entry->IPAddressList == NULL)
+        {
+            ConvertGetAddrInfoAndGetNameInfoErrorsToPal(EAI_MEMORY);
```

Comment:
You're calling this but ignoring its return value?  Did you mean to return here?

Reply:
You're right and thanks for catching this! Fixed.